ports/stm32/boards/NUCLEO_WB55/rfcore_firmware.py

```diff
@@ -0,0 +1,532 @@
+        mem32[stm.FLASH + stm.FLASH_CR] = cr | 1 << 16  # STRT
+        self.wait_not_busy()
+        mem32[stm.FLASH + stm.FLASH_CR] = 0
+
+    def write(self, addr, buf, sz):
+        assert sz % 4 == 0
+        self.wait_not_busy()
+        cr = 1 << 0  # PG
+        mem32[stm.FLASH + stm.FLASH_CR] = cr
+        buf_addr = addressof(buf)
+        off = 0
+        while off < sz:
+            mem32[addr + off] = mem32[buf_addr + off] ^ OBFUSCATION_KEY
+            off += 4
+            if off % 8 == 0:
+                self.wait_not_busy()
+        if off % 8:
+            mem32[addr + off] = 0
+            self.wait_not_busy()
+        mem32[stm.FLASH + stm.FLASH_CR] = 0
+
+
+def _copy_file_to_flash(filename, addr):
+    flash = _Flash()
+    flash.unlock()
+    try:
+        # Erase the entire staging area in flash.
+        erase_addr = FILESYSTEM_END
+        sfr_sfsa = mem32[stm.FLASH + stm.FLASH_SFR] & 0xFF
+        erase_limit = 0x08000000 + sfr_sfsa * 4096
+        while erase_addr < erase_limit:
+            flash.erase_page((erase_addr - 0x08000000) // 4096)
+            erase_addr += 4096
+
+        # Write the contents of the firmware (note flash.write will apply the
+        # XOR de-obfuscation).
+        with open(filename, "rb") as f:
+            buf = bytearray(4096)
+
+            while 1:
+                sz = f.readinto(buf)
+                if sz == 0:
+                    break
+                flash.write(addr, buf, sz)
+                addr += 4096
+
+    finally:
+        flash.lock()
+
+
+def _parse_vendor_response(data):
+    assert len(data) >= 7
+    assert data[0] == HCI_KIND_VENDOR_RESPONSE
+    assert data[1] == 0x0E
+    # assert data[3] == 0xff # "Num HCI" -- docs say 0xff, but we see 0x01
+    op = (data[5] << 8) | data[4]
+    return (op >> 10, op & 0x3FF, data[6], data[7] if len(data) > 7 else 0)
+
+
+def _run_sys_hci_cmd(ogf, ocf, buf=b""):
+    try:
+        ogf_out, ocf_out, status, result = _parse_vendor_response(rfcore.sys_hci(ogf, ocf, buf))
+    except OSError:
+        # Timeout or FUS not active.
+        return (0xFF, 0xFF)
+    assert ogf_out == ogf
+    assert ocf_out == ocf
+    return (status, result)
+
+
+def _fus_get_state():
+    return _run_sys_hci_cmd(OGF_VENDOR, OCF_FUS_GET_STATE)
+
+
+def _fus_is_idle():
+    return _fus_get_state() == (0, 0)
+
+
+def _fus_start_ws():
+    return _run_sys_hci_cmd(OGF_VENDOR, OCF_FUS_START_WS)
+
+
+def _fus_fwdelete():
+    return _run_sys_hci_cmd(OGF_VENDOR, OCF_FUS_FW_DELETE)
+
+
+def _fus_run_fwupgrade(addr):
+    # Note: Address is ignored by the FUS (see comments above).
+    return _run_sys_hci_cmd(OGF_VENDOR, OCF_FUS_FW_UPGRADE, struct.pack("<I", addr))
+
+
+# Get/set current state/reason to RTC Backup Domain. Using the
+# second/third-last register (17, 18) as (although unused on WB55), the last
+# one (19) is reserved for restoring the frequency.
+def _read_state():
+    return machine.mem32[stm.RTC + stm.RTC_BKP18R]
```

Comment:
perhaps make the registers (17/18) constants at the top of the file to easily change them (`_RTC_REG_STATE = const(18)` etc)